### ISSUE TITLE
`needless_return_with_question_mark` false positive when ? converts Error type

### ISSUE THREAD

**Issue by overvenus** [rust-lang/rust-clippy#11982]
### Summary

`needless_return_with_question_mark` false positive when ? converts Error type.

### Lint Name

needless_return_with_question_mark

### Reproducer

I tried this code:

```rust
// https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dd648aea8bffff2237f5e449761918d3

mod bar {
    pub struct Error;

    pub fn foo(_: bool) -> Result<(), Error> {
        Ok(())
    }
}

pub struct Error;

impl From<bar::Error> for Error {
    fn from(_: bar::Error) -> Self {
        Error
    }
}

fn foo(ok: bool) -> Result<(), Error> {
    if !ok {
        return bar::foo(ok).map(|_| Ok(()))?;
    };
    if ok {
        Ok(())
    } else {
        Err(Error)
    }
}

fn main() {
    let _ = foo(true);
}
```

I saw this happen:

```
    Checking playground v0.0.1 (/playground)
warning: unneeded `return` statement with `?` operator
  --> src/main.rs:19:9
   |
19 |         return bar::foo(ok).map(|_| Ok(()))?;
   |         ^^^^^^^ help: remove it
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return_with_question_mark
   = note: `#[warn(clippy::needless_return_with_question_mark)]` on by default

warning: `playground` (bin "playground") generated 1 warning (run `cargo clippy --fix --bin "playground"` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 0.65s
```

I expected to see this happen:

No warning, `?` is necessary for type convertion.

Also, I found that applying `cargo clippy --fix` breaks compilation.
 
```shell
cargo clippy --fix --bin t
    Checking t v0.1.0 (/home/user/t)
warning: failed to automatically apply fixes suggested by rustc to crate `t`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed
  --> src/main.rs:24:30
   |
24 |         bar::foo(ok).map(|_| Ok(()))?;
   |                              ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
   |
help: consider specifying the generic arguments
   |
24 |         bar::foo(ok).map(|_| Ok::<(), E>(()))?;
   |                                +++++++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.
```

### Version

```text
rustc -Vv        
rustc 1.76.0-nightly (06e02d5b2 2023-12-09)
binary: rustc
commit-hash: 06e02d5b259c1e88cbf0c74366d9e0a4c7cfd6d9
commit-date: 2023-12-09
host: x86_64-unknown-linux-gnu
release: 1.76.0-nightly
LLVM version: 17.0.5
```

**Comment by PartiallyTyped** [comment]
The issue doesn't seem to be the need for type conversion, but rather something else, maybe something related to pathing / branching?

This reproduces the issue

```
fn issue11982_2() {
    fn bar(_: bool) -> Result<(), String> {
        Ok(())
    }

    fn foo(ok: bool) -> Result<(), String> {
        if !ok {
            return bar(ok).map(|_| Ok::<(), String>(()))?;
        };
        Ok(())
    }
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1d5b65d0e9d57b5ef57c3bfbd638191b

This is as expected and raises only unnecessary return without question mark warning.

```
mod bar {
    pub struct Error;
    pub fn foo(_: bool) -> Result<(), Error> {
        Ok(())
    }
}

pub struct Error;

impl From<bar::Error> for Error {
    fn from(_: bar::Error) -> Self {
        Error
    }
}

fn foo(ok: bool) -> Result<(), Error> {
    return bar::foo(ok).map(|_| Ok::<(), Error>(()))?;
}
```
